### PR TITLE
Fire function implemented. tank shoots bullet.

### PR DESCRIPTION
**Pull request by peregurin**
when the tank view gets the input to fire the weapon, it calls a function in the controller. The line of communication is:-
Tankview -> tankcont -> TankSer -> BulletSer -> BulletCont -> bulletView.
Right now bullet view has the logic of firing. But this violates the MVC model. The logic needs to be in the BulletCont.
Please suggest changes so that this becomes possible.

### CODE REVIEW

File: Assets/_Scripts/Bullet/Bullet MVC/BulletController.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+        public void FireBullet(Vector3 tankRotation)
+        {
+            BulletView.FireBullet(tankRotation);
+        }
+
+        public void DestroyController()
+        {
+            Debug.Log("bullet controller destroyed...");
+            BulletService.DestroyControllerAndModel();
```
Comment (mayankgrover) [inline Assets/_Scripts/Bullet/Bullet MVC/BulletController.cs]: use singleton

File: Assets/_Scripts/Bullet/Bullet MVC/BulletView.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+            transform.Translate(Vector3.forward * bulletController.BulletModel.Speed * Time.deltaTime);
+
+            CheckBulletBounds();
+
+        }
+
+        private void CheckBulletBounds()
+        {
+            if((Mathf.Abs(transform.position.x) > 48f) || (Mathf.Abs(transform.position.z) > 48f)){
```
Comment (mayankgrover) [inline Assets/_Scripts/Bullet/Bullet MVC/BulletView.cs]: use collider

File: Assets/_Scripts/Tank/Tank MVC/TankController.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        TankService tankService;
+        public TankModel TankModel { get; }
+        public TankView TankView { get; }
+
+        public TankController(TankModel tankModel, TankView[] tankView)
+        {
+            Debug.Log("tank controller created");
+            TankModel = tankModel;
+            TankView = GameObject.Instantiate<TankView>(tankView[(int)TankModel.TankType]); // now here on the basis of what the player has selected as tanktype, need to pass a variable in the tankView array.
```
Comment (mayankgrover) [inline Assets/_Scripts/Tank/Tank MVC/TankController.cs]: dictionary

File: Assets/_Scripts/Singletons/BulletService.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+using Bullet.Model;
+using System;
+
+namespace Bullet.Service
+{
+    public class BulletService : MonoSingletonGeneric<BulletService>
+    {
+        public BulletView BulletView;
+
+        public void CreateNewBullet(Vector3 position, Vector3 tankRotation)
```
Comment (mayankgrover) [inline Assets/_Scripts/Singletons/BulletService.cs]: return BC to tha tank

**Review by mayankgrover — CHANGES_REQUESTED** [review]

**Review by vijit101 — COMMENTED** [review]
- Remove all unnecessary usings 
- Create a function as  public  bulletcontroller GetBullet(Tanktype type) in the bullet service so that anyone can get a bullet when someone needs it and any type of bullet

**Review by vijit101 — COMMENTED** [review]
- Use scriptable objects in the model to transfer the data 
- Add extension functions to use and practice 
- Add enums for bullet types

**Pull request merged**